### DISCUSSION TITLE
V2: 在表单 Schema 变化的情况下，如果想保留填写的 value 要怎么做？

### DISCUSSION THREAD

**Discussion by kucinl** [alibaba/formily#1224] [Q&A 2.x]
比如填写过程中需要删除部分表单项，在 field 变化的情况下，如果不重新调用 createForm 创建新的 form, 提交和校验时都会提交和校验已删除的值；
目前的做法伪代码如下
```
   const formValues = useMemo(() => observable({}), []);

// 在 formItem 变化的时候，重新创建 form 并使用已经记录下的values
   const form = useMemo(() => createForm({
    values: formValues,
  }), [formItemList]);

  const onSubmit = useCallback((values: any) => {
  // 过滤values中已经被删除的值
    return apis.save(doFilter(values, formItemList));
  }, []);
```

感觉不是很科学，不知道有没有更好的办法

**Comment by janryWang** [comment]
可以在Schema变化前把form.values保存起来，变化之后再调用setValues灌到新form上去

**Reply by kucinl** [reply]
不是的，这里有两个问题，第一个问题是，schema变化后，如果不重新调用createForm创建新的form,使用老form的话，在提交校验的时候schema变化前的内容也会被校验（这里我觉得是个 bug);
第二个问题是，自己将 form.values 保存起来后，提交的时候会将schema变化前的value也提交上去，这里必须要自己处理过滤一下才行；
总体上感觉 schema 变化的情况下默认行为有点问题，需要自己处理的有点多

**Reply by janryWang** [reply]
schema变化，form是感知不到的，它内部的字段树是不会自动销毁的，内部所有状态树都还在的，所以你校验是肯定会生效的，1.x就是有自动逻辑，结果会导致很多边界问题，所以到2.0就需要用户手动销毁了

**Reply by elsonz** [reply]
> schema变化，form是感知不到的，它内部的字段树是不会自动销毁的，内部所有状态树都还在的，所以你校验是肯定会生效的，1.x就是有自动逻辑，结果会导致很多边界问题，所以到2.0就需要用户手动销毁了

如何手动销毁字段？文档里找了很久，没有找到相关介绍

**Reply by liuweiGL** [reply]
创建一个新的 form 会有其他的问题吗，为什么一定要去 form 里面销毁呢？

**Reply by liuweiGL** [reply]
手动销毁的话需要查找到 field 的实例，然后调用 onUnmount 方法：
```ts
form.query('xxx').take().onUnmount()
```

**Comment by liuweiGL** [comment]
如果你需要使用 schema 模式，我建议你封装一个自定义的 SchemaForm，把个性化的逻辑都放在组件里面来处理。
比如 schema 变化时自动创建一个新的 form，这个用 useEffect 来实现就很方便。
